File: 8-k-means/k_means.R
dataset = read.csv("Mall_Customers.csv")
x = dataset[,4:5]

#metodo del codo
set.seed(6)
wcss = vector()
for (i in 1:10) {
  wcss[i] <- sum(kmeans(x,i)$withinss)
}
plot(1:10,wcss,type = 'b', main = "Metodo del codo",
     xlab="Numero de clusters",
     ylab="WCSS")

#Aplicar el algoritmo de k-means con k optimo
set.seed(29)
classifier = kmeans(x,5,iter.max = 300,nstart = 10)

ver = classifier$cluster
#visualizacion de los clusters
install.packages("cluster")
library(cluster)
clusplot(x,
         ver,
         lines = 0,
         shade = TRUE,
         color = TRUE,
         labels = 1,
         plotchar = FALSE,
         span = TRUE,
         main = "Clustering de clientes",
         xlab = "ingresos anuales",
         ylab = "Puntuacion")

File: 8-k-means/k_means.py
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

dataset = pd.read_csv("Mall_Customers.csv")
x = dataset.iloc[:,[3,4]].values

#metodo del codo
from sklearn.cluster import KMeans
wcss = []
for i in range(1,11):
    kmeans = KMeans(n_clusters=i,init="k-means++",max_iter=300,n_init = 10, random_state=0)
    kmeans.fit(x)
    wcss.append(kmeans.inertia_)
    
plt.plot(range(1,11),wcss)
plt.title("Metodo del codo")
plt.xlabel("Numero de clusters")
plt.ylabel("WCSS(k)")
plt.show()

#aplicar el metodo de kmeans para segmentar la data
kmeans = KMeans(n_clusters = 5,init="k-means++",max_iter=300,n_init = 10, random_state=0)
y_pred = kmeans.fit_predict(x)

#visualizacion de los clusters
plt.scatter(x[y_pred == 0,0],x[y_pred == 0,1],s=100,c="red",label="Cluster 1")
plt.scatter(x[y_pred == 1,0],x[y_pred == 1,1],s=100,c="blue",label="Cluster 2")
plt.scatter(x[y_pred == 2,0],x[y_pred == 2,1],s=100,c="green",label="Cluster 3")
plt.scatter(x[y_pred == 3,0],x[y_pred == 3,1],s=100,c="brown",label="Cluster 4")
plt.scatter(x[y_pred == 4,0],x[y_pred == 4,1],s=100,c="cyan",label="Cluster 5")
plt.scatter(kmeans.cluster_centers_[:,0],kmeans.cluster_centers_[:,1],s=150,c="yellow",label="Baricentros")
plt.title("Cluster de clientes")
plt.xlabel("ingresos anuales")
plt.ylabel("Puantuacion de gasto")
plt.legend()
plt.show